### PR TITLE
 Fix: False positive on const increment

### PR DESCRIPTION
**Pull request by singalhimanshu**
The following code should not report an error now:
    const a = [42];
    ++a[0];

Fix: #307

### CODE REVIEW

File: src/parse.cpp [hunk 1]
```diff
@@ -248,7 +248,7 @@ expression* parser::parse_primary_expression(precedence prec) {
     source_code_span operator_span = this->peek().span();
     this->skip();
     expression* child = this->parse_expression(
-        precedence{.binary_operators = false, .commas = false});
+        precedence{.binary_operators = true, .commas = false});
```
Comment (strager) [inline src/parse.cpp:251]: This change doesn't seem right. Won't it break code like this:
```js
++x + y

(++x) + y  // before your commit
++(x + y)  // after your commit
```

**Review by strager — CHANGES_REQUESTED** [review]
Blocking: Add tests.

File: src/parse.cpp [hunk 2]
```diff
@@ -791,6 +791,9 @@ expression* parser::parse_expression_remainder(expression* ast,
     if (!prec.math_or_logical_or_assignment) {
```
Comment (strager) [inline src/parse.cpp:791]: Could we reuse `math_or_logical_or_assignment`?

**Review by strager — CHANGES_REQUESTED** [review]
Blocking: Add tests.

File: src/parse.cpp (hunk 2 above)
Comment (singalhimanshu) [reply src/parse.cpp]: I guess we can.

File: src/parse.cpp (hunk 1 above)
Comment (singalhimanshu) [reply src/parse.cpp]: > This change doesn't seem right. Won't it break code like this:

Yup it was breaking.

**Comment by singalhimanshu** [comment]
I have added some tests, let me know if there are still more conditions that are to be tested.

File: test/test-parse-expression.cpp [hunk 3]
```diff
@@ -1419,6 +1419,24 @@ TEST_F(test_parse_expression, parse_prefix_plusplus_minusminus) {
...
+  }
+
+  {
+    test_parser p(u8"--x\n+\ny"_sv);
+
+    expression* ast = p.parse_expression();
+    EXPECT_EQ(summarize(ast), "binary(rwunary(var x), var y)");
+    EXPECT_THAT(p.errors(), IsEmpty());
+  }
+}
```
Comment (strager) [inline test/test-parse-expression.cpp:1438]: I'd like tests for `++x.y` and `++x[y]` in test-parse-express.cpp too.

File: src/parse.cpp [hunk 4]
```diff
@@ -79,7 +79,7 @@ parser::loop_guard parser::enter_loop() {
 
 expression* parser::parse_expression(precedence prec) {
   expression* ast = this->parse_primary_expression(prec);
-  if (!prec.binary_operators) {
+  if (!prec.binary_operators && prec.math_or_logical_or_assignment) {
```
Comment (strager) [inline src/parse.cpp:82]: This looks weird... But it works... 🤷‍♀️

**Review by strager — APPROVED** [review]

**Comment by singalhimanshu** [comment]
I have added the test. I think we can ship it.

**Review by strager — APPROVED** [review]

**Pull request merged**